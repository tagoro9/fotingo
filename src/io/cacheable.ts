import * as Keyv from 'keyv';
import { homedir } from 'os';

const path = `sqlite://${homedir()}/.fotingo_config/cache.sqlite3`;
const keyv = new Keyv(path);

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type PromiseFunction = (...args: any[]) => Promise<any>;

type Cacheable = (
  target: object,
  propertyKey: string | symbol,
  descriptor: TypedPropertyDescriptor<PromiseFunction>,
) => TypedPropertyDescriptor<PromiseFunction>;

/**
 * Decorator that caches the output of the decorated function
 * in an external data source (SQLite DB) so it can be
 * accessed across multiple executions
 * Caching is based on the function input, the specified
 * prefix and the number of minutes the data is supposed to be
 * cached
 */
export function cacheable({
  getPrefix,
  minutes,
}: {
  getPrefix?: () => string;
  minutes: number;
}): Cacheable {
  return (
    target: object,
    propertyKey: string | symbol,
    descriptor: TypedPropertyDescriptor<PromiseFunction>,
  ): TypedPropertyDescriptor<PromiseFunction> => {
    const method = descriptor.value;
    if (typeof method !== 'function') {
      throw new TypeError(
        `@cacheable decorator can only be applied to methods not: ${typeof method}`,
      );
    }

    const cachedFn: PromiseFunction = async function(...args) {
      const prefix = getPrefix ? getPrefix.call(this) : '';
      const key = `${prefix}${target.constructor.name}_${String(propertyKey)}`;
      const cachedValue = await keyv.get(key);
      if (cachedValue) {
        return Promise.resolve(cachedValue);
      }
      const result = await method.call(this, ...args);
      await keyv.set(key, result, minutes * 60 * 1000);
      return Promise.resolve(result);
    };

    descriptor.value = cachedFn;
    return descriptor;
  };
}

// One day in minutes
export const ONE_DAY = 60 * 24;
