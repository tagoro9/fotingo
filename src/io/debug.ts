import createDebugger, { Debugger } from 'debug';

export const debug = createDebugger('fotingo') as Debugger;
