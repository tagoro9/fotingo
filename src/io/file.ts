import * as editor from 'editor';
import * as fs from 'fs';
import { resolve } from 'path';
import { zipObj } from 'ramda';
import * as tmp from 'tmp';
import { promisify } from 'util';

interface EditVirtualFileOptions {
  // Extension the file should have
  extension: string;
  // Initial content of the file
  initialContent?: string;
  // Prefix to include in the file name
  prefix: string;
}

interface FileData {
  // Function to be invoked when done when working with the file
  clean: () => void;
  // File descriptor
  fd: string;
  // Path to tmp file
  path: string;
}

const readFile = promisify(fs.readFile);
const writeFile = promisify(fs.writeFile);

/**
 * Promisified version of tmp.file
 * @param fileOptions File creation options
 */
function createTmpFile(fileOptions: tmp.FileOptions): Promise<FileData> {
  return new Promise((resolve, reject) => {
    tmp.file(fileOptions, (err, ...args) => {
      if (err) {
        return reject(err);
      }
      resolve((zipObj(['path', 'fd', 'clean'], args) as unknown) as FileData);
    });
  });
}

/**
 * Open users preferred editor to edit a file and return the file
 * contents. If the process exits with a non zero code or it is killed
 * by a signal the function will throw an error
 * @param path Path fo the file to edit
 */
async function editFile(path: string): Promise<string> {
  return new Promise((res, reject) => {
    editor(path, (code?: number, signal?: string) => {
      if (code !== 0 || signal !== null) {
        return reject(new Error('The editor exited with an error code'));
      }
      res(readFile(path, 'utf-8'));
    });
  });
}

/**
 * Allow user to input structured content by creating a tmp file and
 * opening their preferred editor. The function resolves the contents of the file
 * @param options Editing options
 */
export async function editVirtualFile(options: EditVirtualFileOptions): Promise<string> {
  const fileData = await createTmpFile({
    postfix: `.${options.extension}`,
    prefix: options.prefix,
  });
  await writeFile(fileData.fd, options.initialContent);
  const fileContent = await editFile(fileData.path);
  fileData.clean();
  return fileContent;
}

/**
 * Get the file content of the file name if it exists in any of the specified folders
 * @param name File name
 * @param root Root folder where to start search
 * @param folders Folders where to look for the file
 */
export async function getFileContent(
  name: string,
  root: string,
  folders: string[],
): Promise<string | undefined> {
  const data = await Promise.all(
    folders
      .map(folder => resolve(root, folder, name))
      .map(p => readFile(p, 'utf-8').catch(() => undefined)),
  );
  return data.filter(e => e !== undefined)[0];
}
