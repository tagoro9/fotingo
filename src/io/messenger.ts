import { boundMethod } from 'autobind-decorator';
import { Observable, Subject } from 'rxjs';
import { take } from 'rxjs/operators';

export enum MessageType {
  ERROR = 'error',
  INFO = 'info',
  REQUEST = 'request',
  STATUS = 'status',
}

export enum RequestType {
  TEXT = 'text',
  PASSWORD = 'password',
  SELECT = 'select',
}

export enum Emoji {
  PENCIL = 'pencil2',
  ARROW_UP = 'arrow_up',
  BUG = 'bug',
  MAG_RIGHT = 'mag_right',
  BOOKS = 'books',
  BOOKMARK = 'books',
  BOOM = 'boom',
  OK = 'ok_hand',
  QUESTION = 'grey_question',
  LINK = 'link',
  SHIP = 'ship',
  TADA = 'tada',
}

export interface Message {
  detail?: string;
  emoji?: Emoji;
  message: string;
  type: MessageType;
  showSpinner: boolean;
}

export interface Request extends Message {
  type: MessageType.REQUEST;
  requestType: RequestType;
}

export interface Status extends Message {
  inThread: boolean;
}

export interface SelectRequest extends Request {
  options: Array<{ label: string; value: any }>;
}

export class Messenger {
  private subject: Subject<Message | Request | Status>;
  private requestSubject: Subject<any>;

  constructor() {
    this.subject = new Subject();
    this.requestSubject = new Subject();
  }

  @boundMethod
  public emit(message: string, emoji?: Emoji): void {
    this.subject.next({ emoji, message, type: MessageType.INFO, showSpinner: true });
  }

  @boundMethod
  public inThread(value: boolean): void {
    this.subject.next({
      inThread: value,
      message: 'It does not matter',
      showSpinner: false,
      type: MessageType.STATUS,
    });
  }

  public send(data: string): void {
    this.requestSubject.next(data);
  }

  public request<T>(
    question: string,
    options: { options?: Array<{ label: string; value: T }>; hint?: string } = {},
  ): Observable<T> {
    const request$ = this.requestSubject.pipe(take(1));
    this.subject.next({
      detail: options.hint,
      message: question,
      requestType: options.options ? RequestType.SELECT : RequestType.TEXT,
      showSpinner: false,
      type: MessageType.REQUEST,
      ...(options.options ? { options: options.options } : {}),
    });
    return request$;
  }

  @boundMethod
  public error(error: Error): void {
    this.subject.error(error);
  }

  public onMessage(fn: (msg: Message) => void): void {
    this.subject.subscribe({
      next: fn,
    });
  }

  public stop(): void {
    this.subject.complete();
  }
}
