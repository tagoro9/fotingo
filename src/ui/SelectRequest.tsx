import React = require('react');

import { Box } from 'ink';
import SelectInput, { Item } from 'ink-select-input';

import { SelectRequestProps } from './props';
import { Question } from './Question';

export function SelectRequest({ onSubmit, request }: SelectRequestProps): JSX.Element {
  return (
    <Box flexDirection="column">
      <Question {...request} />
      <SelectInput
        items={request.options}
        onSelect={(item: Item): void => void onSubmit(String(item.value))}
      />
    </Box>
  );
}
