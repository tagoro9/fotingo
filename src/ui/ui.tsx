import React = require('react');

import { render } from 'ink';

import { Fotingo } from './Fotingo';
import { FotingoProps } from './props';

export function renderUi(props: FotingoProps): void {
  render(<Fotingo {...props} />);
}
