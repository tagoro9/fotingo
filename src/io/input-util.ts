import { take, uniqBy } from 'ramda';

import { Messenger } from './messenger';
import { series } from './promise-util';

interface AskToSelectMatchData<T> {
  data: T[][];
  options: string[];
  useDefaults: boolean;
  getQuestion: (item: string) => string;
  getLabel: (item: T) => string;
  getValue: (item: T) => string;
}

/**
 * Given the options that a user selected and the found matches, ask the user to select
 * the best match out of the first 5 matches. Select the first match in the list if using defaults
 * or the list only has one element
 * @param options Options for selecting the matches
 * @param options.data Data found for the options introduced by the user
 * @param options.getLabel Function that given a match, returns the label to present to the user
 * @param options.getQuestion Function that returns the question to present to the user. It receives the option introduced by the user
 * @param options.getValue Function that given a match, returns its value (typically an id)
 * @param options.options List of options that the use introduced
 * @param options.useDefaults Flag indicating if the useDefaults options was set
 */
export function maybeAskUserToSelectMatches<T>(
  { data, getLabel, getQuestion, getValue, options, useDefaults }: AskToSelectMatchData<T>,
  messenger: Messenger,
): Promise<T[]> {
  return series(
    data.map((matches, i) => () => {
      if (!matches || matches.length === 0) {
        throw new Error(`No match found for ${options[i]}`);
      }
      if (useDefaults || matches.length === 1) {
        return Promise.resolve(matches[0]);
      }
      return messenger
        .request(getQuestion(options[i]), {
          options: uniqBy<T, string>(getValue, take(5, matches)).map(r => ({
            label: getLabel(r),
            value: getValue(r),
          })),
        })
        .toPromise()
        .then((option: string) => matches.find(r => String(getValue(r)) === String(option)));
    }),
  );
}
