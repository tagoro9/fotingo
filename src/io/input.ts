import { take, uniqBy } from 'ramda';
import { series } from 'src/util/promise';

import { Messenger } from './messenger';

interface AskToSelectMatchData<T> {
  data: T[][];
  getLabel: (item: T) => string;
  getQuestion: (item: string) => string;
  getValue: (item: T) => string;
  options: string[];
  useDefaults: boolean;
}

/**
 * Given the options that a user selected and the found matches, ask the user to select
 * the best match out of the first 5 matches. Select the first match in the list if using defaults
 * or the list only has one element
 * @param options Options for selecting the matches
 * @param options.data Data found for the options introduced by the user
 * @param options.getLabel Function that given a match, returns the label to present to the user
 * @param options.getQuestion Function that returns the question to present to the user. It receives the option introduced by the user
 * @param options.getValue Function that given a match, returns its value (typically an id)
 * @param options.options List of options that the use introduced
 * @param options.useDefaults Flag indicating if the useDefaults options was set
 */
export function maybeAskUserToSelectMatches<T>(
  { data, getLabel, getQuestion, getValue, options, useDefaults }: AskToSelectMatchData<T>,
  messenger: Messenger,
): Promise<T[]> {
  return series(
    data.map((matches, i) => (): Promise<T> => {
      if (!matches || matches.length === 0) {
        throw new Error(`No match found for ${options[i]}`);
      }
      if (useDefaults || matches.length === 1) {
        return Promise.resolve(matches[0]);
      }
      return (
        messenger
          .request(getQuestion(options[i]), {
            options: uniqBy<T, string>(getValue, take(5, matches)).map(r => ({
              label: getLabel(r),
              value: getValue(r),
            })),
          })
          .toPromise()
          // We know the user selected an option
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          .then((option: string) => matches.find(r => String(getValue(r)) === String(option))!)
      );
    }),
  );
}
