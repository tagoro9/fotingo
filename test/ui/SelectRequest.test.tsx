import 'jest';

import { render } from 'ink-testing-library';
import { MessageType, RequestType, SelectRequest as SelectRequestType } from 'src/io/messenger';
import { SelectRequest } from 'src/ui/SelectRequest';

import React = require('react');

describe('<SelectRequest />', () => {
  const request: SelectRequestType = {
    allowTextSearch: true,
    message: 'Select an option',
    options: [
      { label: 'Item 1', value: 1 },
      { label: 'Item 2', value: 2 },
      { label: 'Item 3', value: 3 },
    ],
    showSpinner: false,
    requestType: RequestType.SELECT,
    type: MessageType.REQUEST,
  };

  test('displays the list of options filtering by the text input', async () => {
    const actual = render(<SelectRequest onSubmit={jest.fn()} request={request} />);
    expect(actual.lastFrame()).toMatchSnapshot();
    actual.stdin.write('1');
    expect(actual.lastFrame()).toMatchSnapshot();
  });

  // eslint-disable-next-line jest/no-disabled-tests
  test.skip('allows to navigate the list of options using the arrow keys', async () => {
    const ARROW_DOWN = '\u001B[B';

    const actual = render(
      <SelectRequest
        onSubmit={jest.fn()}
        request={{
          ...request,
          allowTextSearch: false,
        }}
      />,
    );
    expect(actual.lastFrame()).toMatchSnapshot();
    actual.stdin.write(ARROW_DOWN);
    expect(actual.lastFrame()).toMatchSnapshot();
  });
});
