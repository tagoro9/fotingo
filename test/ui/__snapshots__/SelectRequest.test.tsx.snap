// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectRequest /> allows to navigate the list of options using the arrow keys 1`] = `
"‚ùî   Select an option

 [34m‚ùØ[39m [34mItem 1[39m
   Item 2
   Item 3
"
`;

exports[`<SelectRequest /> allows to navigate the list of options using the arrow keys 2`] = `
"‚ùî   Select an option

 [34m‚ùØ[39m [34mItem 1[39m
   Item 2
   Item 3
"
`;

exports[`<SelectRequest /> displays the list of options filtering by the text input 1`] = `
"‚ùî   Select an option

 ‚ùØ Item 1
   Item 2
   Item 3
"
`;

exports[`<SelectRequest /> displays the list of options filtering by the text input 2`] = `
"‚ùî   Select an option 1

 ‚ùØ Item 1
"
`;
